design/hostclaim-multitenancy.md

```diff
@@ -0,0 +1,582 @@
+  to a tenant, similar to how quotas work for pods. However, the specification
+  of such a framework will be addressed in another design document.
+* Implementing network isolation between tenant clusters using a common set
+  of BareMetalHosts.
+
+## Proposal
+
+### User Stories
+
+#### Deployment of Simple Workloads
+
+##### User Point of View
+
+As an *end-user*, I would like to execute a workload on an arbitrary server.
+
+The OS image is available in qcow format on a remote server at ``url_image``.
+It supports cloud-init and a script can launch the workload at boot time
+(e.g., a systemd service).
+
+The cluster offers bare-metal as a service using Metal3 baremetal-operator.
+However, as a regular user, I am not allowed to directly access the definitions
+of the servers, but I can read the result of the inspection (HardwareData).
+The HardwareData resource also contains a copy of the metadata of the
+BareMetalHost visible to the end-user.
+
+All HardwareData are labeled with an ``infra-kind`` label whose
+value depends on the characteristics of the computer.
+
+* I can read HardwareData in namespace ``infra``. I can find which selector
+  I should use to find a machine with the right hardware details.
+* I create a resource with the following content:
+
+  ```yaml
+  apiVersion: metal3.io/v1alpha1
+  kind: HostClaim
+  metadata:
+    name: my-host
+  spec:
+    online: false
+
+    hostSelector:
+      matchLabels:
+        infra-kind: medium
+  ```
+
+* After a while, the system associates the claim with a real server, and
+  the resource's status is populated with the following information:
+
+  ```yaml
+  status:
+    bareMetalHost:
+      name: server-123
+      namespace: infra
+    conditions:
+    - lastTransitionTime: "2024-03-29T14:33:19Z"
+      status: "True"
+      type: Ready
+    - lastTransitionTime: "2024-03-29T14:33:19Z"
+      status: "True"
+      type: AssociateBMH
+    lastUpdated: "2024-03-29T14:33:19Z"
+  ```
+
+* I create three secrets in the same namespace ``my-user-data``,
+  ``my-metadata``, and ``my-network-data``. I use the information from the
+  status and meta data to customize the scripts they contain.
+* I modify the HostClaim to point to those secrets and start the server:
+
+  ```yaml
+  apiVersion: metal3.io/v1alpha1
+  kind: HostClaim
+  metadata:
+    name: my-host
+  spec:
+    online: true
+    image:
+      checksum: https://url_image.qcow2.md5
+      url: https://url_image.qcow2
+      format: qcow2
+    userData:
+      name: my-user-data
+    networkData:
+      name: my-network-data
+    hostSelector:
+      matchLabels:
+        infra-kind: medium
+  ```
+
+* The workload is launched. When the machine is fully provisioned, the boolean
+  field ready in the status becomes true. I can stop the server by changing
+  the online status. I can also perform a reboot by targeting specific
+  annotations in the reserved ``host.metal3.io`` domain.
```

Comment:
Why not simply accept the same reboot annotation that BMH accepts (and copy it over)?

Reply:
Good catch. host.metal3.io made sense when the HostClaim resource was used to expose BMH metadata. So some labels/annotations were flowing upward from BMH to HostClaims whereas specific annotations like the one for reboot were flowing downward. Now that HardwareData are ivisible and expose metadata, all this complexity is obosolete.